File: kitten-house-care/src/main/java/org/jesperancinha/housing/repository/OwnerRepositoryImpl.java
package org.jesperancinha.housing.repository;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.jesperancinha.housing.model.Owner;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

@Repository
public class OwnerRepositoryImpl implements OwnerRepository {

    private final Owner[] owners;

    public OwnerRepositoryImpl(ObjectMapper objectMapper) throws IOException {
        this.owners = objectMapper.readValue(getClass().getResourceAsStream("/owners.json"), Owner[].class);
    }

    @Override
    public Mono<Owner> getOwnerById(Long id) {
        return Mono.fromCallable(
            () -> Arrays.stream(owners).filter(owner -> owner.getId().equals(id)).findFirst().orElse(null));
    }

    @Override
    public Mono<List<Owner>> getOwnersByIds(List<Long> formerOwners) {
        return Mono.fromCallable(() -> getOwners(formerOwners));
    }

    @Override
    public Mono<String> checkLiability(String address) {
        return Flux
            .fromStream(() -> Arrays.stream(owners).filter(owner -> owner.getAddress().equalsIgnoreCase(address)))
            .reduce((owner, owner2) -> {
                if (owner.getRating() > owner2.getRating()) {
                    return owner2;
                }
                return owner;
            }).map(owner -> owner.getRating() <= 0 ?
                "NOK" :
                "OK");
    }

    @Override
    public List<Owner> getOwnersByIdsNonReactive(List<Long> formerOwners) {
        return getOwners(formerOwners);
    }

    private List<Owner> getOwners(List<Long> formerOwners) {
        return Arrays.stream(owners)
            .filter(owner -> formerOwners.contains(owner.getId()))
            .collect(Collectors.toList());
    }
}

File: kitten-house-care/src/main/kotlin/org/jesperancinha/housing/repository/OwnerRepositoryImpl.kt
package org.jesperancinha.housing.repository

import com.fasterxml.jackson.databind.ObjectMapper
import org.jesperancinha.housing.model.Owner
import org.springframework.stereotype.Repository
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.util.*
import java.util.stream.Collectors

@Repository
class OwnerRepositoryImpl(objectMapper: ObjectMapper) {
    private val owners: Array<Owner?>

    init {
        owners = objectMapper.readValue(javaClass.getResourceAsStream("/owners.json"), Array<Owner>::class.java)
    }

    fun getOwnerById(id: Long): Mono<Owner?> {
        return Mono.fromCallable {
            Arrays.stream(owners).filter { owner: Owner? -> owner.getId() == id }
                .findFirst().orElse(null)
        }
    }

    fun getOwnersByIds(formerOwners: List<Long>): Mono<List<Owner?>> {
        return Mono.fromCallable { getOwners(formerOwners) }
    }

    fun checkLiability(address: String?): Mono<String> {
        return Flux
            .fromStream {
                Arrays.stream(owners).filter { owner: Owner? -> owner.getAddress().equals(address, ignoreCase = true) }
            }
            .reduce { owner: Owner?, owner2: Owner? ->
                if (owner.getRating() > owner2.getRating()) {
                    return@reduce owner2
                }
                owner
            }.map { owner: Owner? -> if (owner.getRating() <= 0) "NOK" else "OK" }
    }

    fun getOwnersByIdsNonReactive(formerOwners: List<Long>): List<Owner?> {
        return getOwners(formerOwners)
    }

    private fun getOwners(formerOwners: List<Long>): List<Owner?> {
        return Arrays.stream(owners)
            .filter { owner: Owner? -> formerOwners.contains(owner.getId()) }
            .collect(Collectors.toList())
    }
}